src/stores/right-panel/RightPanelStore.ts

```diff
@@ -0,0 +1,334 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { logger } from "matrix-js-sdk/src/logger";
+import { EventSubscription } from 'fbemitter';
+
+import defaultDispatcher from '../../dispatcher/dispatcher';
+import { pendingVerificationRequestForUser } from '../../verification';
+import SettingsStore from "../../settings/SettingsStore";
+import { RightPanelPhases/*, RIGHT_PANEL_PHASES_NO_ARGS */ } from "./RightPanelStorePhases";
+import { ActionPayload } from "../../dispatcher/payloads";
+import { Action } from '../../dispatcher/actions';
+import { SettingLevel } from "../../settings/SettingLevel";
+import { UPDATE_EVENT } from '../AsyncStore';
+import { ReadyWatchingStore } from '../ReadyWatchingStore';
+import RoomViewStore from '../RoomViewStore';
+import { IPhaseAndState, IPanelState, convertToStateRoom, convertToStoreRoom } from './RightPanelStoreIPanelState';
+
+const GROUP_PHASES = [
+    RightPanelPhases.GroupMemberList,
+    RightPanelPhases.GroupRoomList,
+    RightPanelPhases.GroupRoomInfo,
+    RightPanelPhases.GroupMemberInfo,
+];
+
+const MEMBER_INFO_PHASES = [
+    RightPanelPhases.RoomMemberInfo,
+    RightPanelPhases.Room3pidMemberInfo,
+    RightPanelPhases.EncryptionPanel,
+];
+
+/**
+ * A class for tracking the state of the right panel between layouts and
+ * sessions. This state includes a history for each room. Each history element contains
+ * the phase (e.g. RoomMemberInfo) and the state (e.g. the member) associated with it.
+ *
+ * Groups are treated the same then rooms (They are also stored in the byRoom object).
+ * This is possible since the store keeps track of the opened room/group -> the store will
+ * provide the correct history for that group/room.
+ */
+export default class RightPanelStore extends ReadyWatchingStore {
+    private static internalInstance: RightPanelStore;
+    private viewedRoomId: string;
+    private isViewingRoom?: boolean;
+    private dispatcherRefRightPanelStore: string;
+    private roomStoreToken: EventSubscription;
+
+    private byRoom: {
+        [roomId: string]: {
+            isOpen: boolean;
+            history: Array<IPhaseAndState>;
+        };
+    } = {};
+    private global?: {
+        isOpen: boolean;
+        history: Array<IPhaseAndState>;
+    } = null;
+
+    private constructor() {
+        super(defaultDispatcher);
+        this.dispatcherRefRightPanelStore = this.dispatcher.register(this.onDispatch);
+    }
+
+    protected async onReady(): Promise<any> {
+        this.roomStoreToken = RoomViewStore.addListener(this.onRoomViewStoreUpdate);
```

Comment:
It's a bit odd to add a listener that does nothing... Perhaps comment this out and mention it should be enabled after removing groups.